### DISCUSSION TITLE
Are there any API Docs / Examples?

### DISCUSSION THREAD

**Discussion by autolyticus** [toeverything/AFFiNE#6052] [Q&A]
Hi! Affine is an amazing project.

I understand this is probably not a priority for you guys right now, but I'm finding it really hard to understand how I can append content to my "Journal" note using the Affine GraphQL API. I'm not that familiar with GraphQL.

I did find that https://app.affine.pro/graphql is there and that the schema and the boilerplate docs are there, but how would I be able to query and get a particular note for basic operations like this?

**Comment by alexfacciorusso** [comment]
Hi team! Would it be possible to let us know more about the API? This is an incredible functionality and we'd love to add automations to Affine!

**Comment by smosgasbord** [comment]
Please!!

**Comment by ddean6232** [comment]
Please respond!?

**Comment by magi-wang** [comment]
Please ~~

**Comment by hieutrluu** [comment]
Please !!!!

**Comment by deg-jc** [comment]
Hello there!

**Comment by PeterWolfframm** [comment]
Hi peeps - would also be very interested in this feature. Any updates?

**Comment by phifa** [comment]
any news here?

**Reply by ddean6232** [reply]
Given up!

**Comment by OliverRM** [comment]
Are there any API Docs / Examples? I'd love if there were!

In lack of official documentation of the API, I had a chat with Copilot and got it to put together rough guide, hopefully helpful to anyone out there puzzled of how to connect to your instance. As a small warning, this guide is result of an AI analysis and I guarantee neither for completeness nor for correctness. But to anyone who has something to add: Don't hesitate to share!

For the future I hope for an official documentation.

---

# AFFiNE Server API Overview (GraphQL + REST + WebSocket Sync + CRDT/Yjs)

This document gives a “layered” overview of the AFFiNE backend API surface based on the server implementation in commit:

- Repo: `toeverything/AFFiNE`
- Commit: `0849b342fa231160dd52fe87ceef12d71835819c`

The server is a NestJS application that exposes multiple API layers:

1. **GraphQL** (primary structured API)
2. **REST / HTTP Controllers** (files, auth, webhooks, utility endpoints)
3. **WebSocket (Socket.IO) “sync”** (collaboration & document updates)
4. **CRDT payload format (Yjs updates)** (the binary content moved across REST + WebSocket)

---

## Base URL + global prefix

The server can be deployed under a subpath.

- `config.server.path` is used as Nest’s global prefix:
  - Source: `packages/backend/server/src/server.ts`
  - `if (config.server.path) app.setGlobalPrefix(config.server.path);`
  - Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/server.ts

Configuration definition:

- Source: `packages/backend/server/src/core/config/config.ts`
- `server.path` is bound to env `AFFINE_SERVER_SUB_PATH`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/config/config.ts

**Implication:** All HTTP routes and GraphQL routes that use `useGlobalPrefix: true` are effectively:

- `/<server.path>/<route>`

---

## 1) GraphQL layer

### Endpoint
GraphQL is mounted at:

- `POST /graphql` (and GraphiQL in development)
- It uses the global prefix (`useGlobalPrefix: true`), so actual path is:
  - `/<server.path>/graphql`

GraphQL module configuration:

- Source: `packages/backend/server/src/base/graphql/index.ts`
- Key lines:
  - `useGlobalPrefix: true`
  - `path: '/graphql'`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/base/graphql/index.ts

### How to dig deeper (GraphQL)
- Search for `@Resolver`, `@Query`, `@Mutation`, `@Subscription` in `packages/backend/server/src`.
- The generated schema file in this commit exists at:
  - `packages/backend/server/src/schema.gql`
  - Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/schema.gql

---

## 2) REST / HTTP controller layer

REST endpoints are implemented via NestJS controllers and are used for:
- file/binary responses (docs, blobs, avatars, attachments)
- auth flows
- telemetry ingestion (HTTP)
- setup/self-host flows
- payment provider webhooks, etc.

### How to dig deeper (REST)
Search for `@Controller` in `packages/backend/server/src` to find all HTTP routes.

GitHub code search example:
- Query: `repo:toeverything/AFFiNE path:packages/backend/server/src @Controller`
- URL: https://github.com/search?q=repo%3Atoeverything%2FAFFiNE+path%3Apackages%2Fbackend%2Fserver%2Fsrc+%40Controller&type=code

### Example REST endpoints (from this commit)

#### Server info
- `GET /info`
- Source: `packages/backend/server/src/app.controller.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/app.controller.ts

#### Doc + blob endpoints (important for CRDT/document layer)
All under the controller:

- Base: `/api/workspaces`
- Source: `packages/backend/server/src/core/workspaces/controller.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/workspaces/controller.ts

Key routes:
- `GET /api/workspaces/:id/docs/:guid` (doc binary)
- `GET /api/workspaces/:id/public-docs/:docId` (public doc binary)
- `GET /api/workspaces/:id/docs/:guid/histories/:timestamp` (history binary)
- `GET /api/workspaces/:id/blobs/:name` (workspace blob binary / redirect to signed URL)
- `GET /api/workspaces/:id/docs/:docId/comment-attachments/:key` (comment attachment binary)

---

## 3) WebSocket layer (Socket.IO)

AFFiNE uses **Socket.IO** (Engine.IO + Socket.IO framing), integrated via NestJS gateways.

In `server.ts`, the server installs a Socket.IO adapter:

- `const adapter = new SocketIoAdapter(app);`
- `app.useWebSocketAdapter(adapter);`
- Source: `packages/backend/server/src/server.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/server.ts

### WebSocket endpoint URL
Because this is Socket.IO, clients connect to the usual Socket.IO engine path:

- Default Engine.IO path: `/socket.io/`
- Combined with global prefix:
  - `/<server.path>/socket.io/`

The custom adapter does not set a custom path; it configures CORS and optional auth middleware.

- Source: `packages/backend/server/src/base/websocket/adapter.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/base/websocket/adapter.ts

### How to dig deeper (WebSocket)
Search for Nest gateways and events:
- `@WebSocketGateway()`
- `@SubscribeMessage('...')`

GitHub code search examples:
- Query: `repo:toeverything/AFFiNE path:packages/backend/server/src @WebSocketGateway`
- Query: `repo:toeverything/AFFiNE path:packages/backend/server/src @SubscribeMessage`
- URL: https://github.com/search?q=repo%3Atoeverything%2FAFFiNE+path%3Apackages%2Fbackend%2Fserver%2Fsrc+%40SubscribeMessage&type=code

### Main sync protocol gateway (docs)
The collaboration + doc syncing protocol is implemented by:

- `SpaceSyncGateway`
- Source: `packages/backend/server/src/core/sync/gateway.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/sync/gateway.ts

Important events (from that file):
- `space:join` / `space:leave`
- `space:load-doc` (diff based on state vector)
- `space:push-doc-update` (write/update)
- `space:broadcast-doc-update` (legacy protocol room)
- `space:broadcast-doc-updates` (current protocol room)
- awareness events: `space:join-awareness`, `space:update-awareness`, etc.

### Internal server-to-server namespace
There is also an internal socket.io namespace:

- `namespace: 's2s'`
- It disconnects clients (internal usage only)
- Source: `packages/backend/server/src/base/event/eventbus.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/base/event/eventbus.ts

---

## 4) CRDT/Yjs layer (binary payloads)

Document content is represented as **Yjs updates** (CRDT deltas). These are transported as:

- **Raw bytes** over REST doc read endpoints (application/octet-stream)
- **Base64-encoded strings** inside Socket.IO JSON messages (sync protocol)

The REST doc endpoint explicitly responds with:
- `content-type: application/octet-stream`
- raw `binResponse.bin`

- Source: `packages/backend/server/src/core/workspaces/controller.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/workspaces/controller.ts

On the Socket.IO side:
- doc updates are base64 in message fields like `update` / `updates[]`
- The gateway encodes `Uint8Array` updates to base64 before broadcasting:
  - `encodeUpdates(updates: Uint8Array[]) => base64[]`
- Source: `packages/backend/server/src/core/sync/gateway.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/sync/gateway.ts

**Important:** Yjs provides the CRDT mechanics, but AFFiNE’s *semantic* document model (which keys/maps represent blocks, titles, etc.) is application-specific. Reading “meaningful text” requires understanding AFFiNE’s schema on top of Yjs.

---

# Common operations (brief examples)

## A) Read a document (REST)
This fetches the doc binary as bytes. The payload is a Yjs update (or merged update), not JSON.

**HTTP:**
- `GET /api/workspaces/:workspaceId/docs/:docId`
- For the **workspace doc** specifically: `docId == workspaceId`

**Node.js example (read bytes + apply to Y.Doc):**
```ts
import * as Y from 'yjs';

async function readDoc(baseUrl: string, workspaceId: string, docId: string) {
  const url = `${baseUrl}/api/workspaces/${workspaceId}/docs/${docId}`;

  const res = await fetch(url, {
    // if auth uses cookies:
    // credentials: 'include',
    headers: {
      // attach auth headers/cookies as appropriate for your deployment
    },
  });

  if (!res.ok) throw new Error(`${res.status} ${res.statusText}`);

  const bytes = new Uint8Array(await res.arrayBuffer());
  const ydoc = new Y.Doc();
  Y.applyUpdate(ydoc, bytes);

  return ydoc;
}
```

Route implementation:
- `GET /api/workspaces/:id/docs/:guid`
- Source: `packages/backend/server/src/core/workspaces/controller.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/workspaces/controller.ts

---

## B) Write/update a document (Socket.IO sync)
Writing is done by pushing a Yjs update over Socket.IO.

Key event for writes:
- `space:push-doc-update`
- Payload: `{ spaceType, spaceId, docId, update: base64(yjsUpdate) }`
- Source: `packages/backend/server/src/core/sync/gateway.ts`
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/sync/gateway.ts

**High-level steps:**
1. connect to Socket.IO at `/<server.path>/socket.io/`
2. `emit('space:join', { spaceType:'workspace', spaceId:<workspaceId>, clientVersion:<semver> })`
3. generate Yjs updates locally (listen to `ydoc.on('update', ...)`)
4. `emit('space:push-doc-update', { spaceType:'workspace', spaceId:<workspaceId>, docId:<docId>, update:<base64> })`

**Node.js example (socket.io-client + Yjs update):**
```ts
import { io } from 'socket.io-client';
import * as Y from 'yjs';

function toBase64(u8: Uint8Array) {
  return Buffer.from(u8).toString('base64');
}

async function pushDocUpdate({
  baseUrl,
  workspaceId,
  docId,
  cookieHeader, // optional: pass "cookie" header if you auth via cookies
}: {
  baseUrl: string;
  workspaceId: string;
  docId: string;
  cookieHeader?: string;
}) {
  // Socket.IO endpoint is baseUrl + (globalPrefix) + /socket.io
  const socket = io(baseUrl, {
    transports: ['websocket'],
    withCredentials: true,
    extraHeaders: cookieHeader ? { cookie: cookieHeader } : undefined,
  });

  await new Promise<void>((resolve, reject) => {
    socket.on('connect', () => resolve());
    socket.on('connect_error', reject);
  });

  // Join space (required; server uses this for membership checks and rooms)
  const joinResp = await socket.emitWithAck('space:join', {
    spaceType: 'workspace',
    spaceId: workspaceId,
    clientVersion: '0.26.0', // must satisfy server's MIN_WS_CLIENT_VERSION
  });
  if (!joinResp?.data?.success) throw new Error('space:join failed');

  // Build a Yjs update locally
  const ydoc = new Y.Doc();
  let updateBytes: Uint8Array | null = null;
  ydoc.on('update', u => (updateBytes = u));

  // NOTE: This mutation is just an example; real AFFiNE docs have app-specific Yjs structure.
  ydoc.getText('content').insert(0, 'hello');

  if (!updateBytes) throw new Error('No Yjs update generated');

  const resp = await socket.emitWithAck('space:push-doc-update', {
    spaceType: 'workspace',
    spaceId: workspaceId,
    docId,
    update: toBase64(updateBytes),
  });

  socket.disconnect();
  return resp;
}
```

---

## C) Load doc diff over Socket.IO (preferred for sync clients)
Instead of downloading the whole doc via REST, sync clients usually do:

- `space:load-doc` with optional `stateVector` (base64), receiving `{missing, state, timestamp}`

Implementation:
- Source: `packages/backend/server/src/core/sync/gateway.ts` (`@SubscribeMessage('space:load-doc')`)
- Reference: https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/sync/gateway.ts

---

# Tips for exploring the API further

## REST discovery
- Search for `@Controller` in `packages/backend/server/src`
- Then inspect each controller’s `@Get`, `@Post`, `@Put`, `@Delete`, etc.
- Tip: many AFFiNE “file-ish” endpoints exist because GraphQL can’t represent raw file streams.

## WebSocket discovery
- Search for `@WebSocketGateway` to find gateways.
- Search for `@SubscribeMessage('...')` to find all events.
- Pay attention to:
  - message interfaces (payload shape)
  - base64 encoding (binary fields)
  - rooms (`sync-025` vs `sync-026`) for protocol compatibility

## GraphQL discovery
- `packages/backend/server/src/schema.gql` shows the schema snapshot for this commit.
- Search for `@Resolver()` and related GraphQL decorators for implementation details.

## Document internals (Yjs schema)
- “Yjs update bytes” are just the transport format.
- To extract meaningful content (blocks, titles, etc.), you must understand AFFiNE’s Yjs data model.
- Start by searching for code that reads/writes Yjs types (e.g. `new Y.Doc()`, `getMap`, `getArray`, `getText`, etc.) in the repo.

---

# Appendix: key source references

- Server bootstrap + global prefix + Socket.IO adapter:
  - `packages/backend/server/src/server.ts`
  - https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/server.ts

- GraphQL endpoint config (`/graphql`):
  - `packages/backend/server/src/base/graphql/index.ts`
  - https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/base/graphql/index.ts

- REST doc/blob endpoints:
  - `packages/backend/server/src/core/workspaces/controller.ts`
  - https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/workspaces/controller.ts

- Sync protocol (Socket.IO events, doc push/load, broadcast):
  - `packages/backend/server/src/core/sync/gateway.ts`
  - https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/sync/gateway.ts

- Socket.IO adapter (CORS + optional auth + redis adapter):
  - `packages/backend/server/src/base/websocket/adapter.ts`
  - https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/base/websocket/adapter.ts

- Server config schema (`AFFINE_SERVER_SUB_PATH`, etc.):
  - `packages/backend/server/src/core/config/config.ts`
  - https://github.com/toeverything/AFFiNE/blob/0849b342fa231160dd52fe87ceef12d71835819c/packages/backend/server/src/core/config/config.ts
